addons/mail/models/mail_link_preview.py

```diff
@@ -0,0 +1,160 @@
+    og_image = fields.Char('image')
+    og_description = fields.Text('description')
+    og_mimetype = fields.Char('mimetype')
+    image_mimetype = fields.Char('image_mimetype')
+
+    # Some websites are blocking non browser user agent.
+    request_headers = {
+        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; rv:91.0) Gecko/20100101 Firefox/91.0'
+    }
+
+    @api.model
+    def _create_link_previews(self, message):
+        tree = html.fromstring(message.body)
+        urls = tree.xpath('//a/@href')
+        link_previews = self.env['mail.link.preview']
+        for url in set(urls):
+            if (len(link_previews) >= 5):
+                break
+            link_previews |= self.env['mail.link.preview'].sudo()._create_link_preview(url, message.id)
+        if not link_previews:
+            return
+        guest = self.env['mail.guest']._get_guest_from_context()
+        if message.model == 'mail.channel' and message.res_id:
+            target = self.env['mail.channel'].browse(message.res_id)
+        elif self.env.user._is_public() and guest:
+            target = guest
+        else:
+            target = self.env.user.partner_id
+        self.env['bus.bus']._sendmany([(target, 'mail.link.preview/insert', link_previews._link_preview_format())])
+
+    @api.model
+    def _create_link_preview(self, url, message_id):
+        if self._is_domain_throttled(url):
+            return self.env['mail.link.preview']
+        link_preview_data = self._get_link_preview_from_url(url)
+        if link_preview_data:
+            link_preview_data['message_id'] = message_id
+            return self.sudo().create(link_preview_data)
+        return self.env['mail.link.preview']
+
+    def _delete_and_notify(self):
+        notifications = []
+        guest = self.env['mail.guest']._get_guest_from_context()
+        for link_preview in self:
+            if link_preview.message_id.model == 'mail.channel' and link_preview.message_id.res_id:
+                target = self.env['mail.channel'].browse(link_preview.message_id.res_id)
+            elif self.env.user._is_public() and guest:
+                target = guest
+            else:
+                target = self.env.user.partner_id
+            notifications.append((target, 'mail.link.preview/delete', {'id': link_preview.id}))
+        self.env['bus.bus']._sendmany(notifications)
+        self.unlink()
+
+    @api.model
+    def _is_link_preview_enabled(self):
+        link_preview_throttle = int(self.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99))
+        return link_preview_throttle > 0
+
+    @api.model
+    def _is_domain_throttled(self, url):
+        domain = urlparse(url).netloc
+        date_interval = fields.Datetime.to_string((datetime.now() - relativedelta(seconds=10)))
+        call_counter = self.search_count([
+            ('source_url', 'ilike', '%' + domain + '%'),
+            ('create_date', '>', date_interval),
```

Comment:
Would probably be a good idea to index on `create_date` if we're going to search on it possibly multiple times per message.

Reply:
Done !